dll/win32/msgina/shutdown.c

```diff
@@ -637,72 +727,65 @@ CreateToolTipForButtons(
-
-    /* If the sleep button has been already replaced earlier, bring sleep button back to its original position */
-    if (bIsSleepButtonReplaced)
+    if (pContext->bTimer)
     {
-        destID = IDC_BUTTON_HIBERNATE;
-        targetedID = IDC_BUTTON_SLEEP;
+        KillTimer(hDlg, pContext->iTimer);
     }
 
-    hwndDest = GetDlgItem(hDlg, destID);
-    hwndTarget = GetDlgItem(hDlg, targetedID);
-
-    /* Get the position of the destination button */
-    GetWindowRect(hwndDest, &rect);
-
-    /* Get the corrected translated coordinates which is relative to the client window */
-    MapWindowPoints(HWND_DESKTOP, hDlg, (LPPOINT)&rect, sizeof(RECT)/sizeof(POINT));
-
-    /* Set the position of targeted button and hide the destination button */
-    SetWindowPos(hwndTarget,
-                 HWND_TOP,
-                 rect.left, rect.top,
-                 0, 0,
-                 SWP_NOSIZE | SWP_NOZORDER | SWP_NOACTIVATE);
-
-    EnableWindow(hwndDest, FALSE);
-    ShowWindow(hwndDest, SW_HIDE);
-    EnableWindow(hwndTarget, TRUE);
-    ShowWindow(hwndTarget, SW_SHOW);
-    SetFocus(hwndTarget);
-
-    if (bIsAltKeyPressed)
+    for (int i = 0; i < DLG_BITMAP_CUSTOMIZABLE; i++)
     {
-        if (!bIsSleepButtonReplaced)
+        if (pContext->hTheme[i])
         {
-            GetDlgItemTextW(hDlg, IDC_BUTTON_HIBERNATE, szBuffer, _countof(szBuffer));
-            SetDlgItemTextW(hDlg, IDC_SLEEP_STATIC, szBuffer);
+            CloseThemeData(pContext->hTheme[i]);
         }
         else
         {
-            GetDlgItemTextW(hDlg, IDC_BUTTON_SLEEP, szBuffer, _countof(szBuffer));
-            SetDlgItemTextW(hDlg, IDC_SLEEP_STATIC, szBuffer);
+            DeleteObject(pContext->hBitmap[i]);
         }
     }
+
+    if (pContext->hTheme[DLG_STATIC_TITLE])
+    {
+        CloseThemeData(pContext->hTheme[DLG_STATIC_TITLE]);
+    }
     else
     {
-        if (!bIsSleepButtonReplaced)
-        {
-            LoadStringW(hInstance, IDS_SHUTDOWN_HIBERNATE, szBuffer, _countof(szBuffer));
-            SetDlgItemTextW(hDlg, IDC_SLEEP_STATIC, szBuffer);
-        }
-        else
-        {
-            LoadStringW(hInstance, IDS_SHUTDOWN_SLEEP, szBuffer, _countof(szBuffer));
-            SetDlgItemTextW(hDlg, IDC_SLEEP_STATIC, szBuffer);
-        }
+        DeleteObject(pContext->hfFont);
+    }
+
+    if (pContext->hTheme[DLG_STATIC_BUTTONS])
+    {
+        CloseThemeData(pContext->hTheme[DLG_STATIC_BUTTONS]);
+    }
+
+    /* Remove the subclass from the buttons */
+    for (int i = 0; i < NUMBER_OF_BUTTONS; i++)
+    {
+        SetWindowLongPtrW(GetDlgItem(hDlg, IDC_BUTTON_SHUTDOWN + i), GWLP_WNDPROC, (LONG_PTR)pContext->OldButtonProc);
+    }
+}
+
+VOID OnTimer(
+    HWND hDlg,
+    PSHUTDOWN_DLG_CONTEXT pContext)
+{
+    BOOL ReplaceButton = !!(GetKeyState(VK_SHIFT) & 0x8000);
+
+    if (ReplaceButton && !pContext->bIsSleepButtonReplaced)
+    {
+        ChangeRequiredButton(hDlg,
+                             pContext);
```

Comment:
Parameter could be placed on the same line as hDlg (same comment for the other function call below too).